### ISSUE TITLE
sklearn快速入门

### ISSUE THREAD

**Issue by Shengqi-Pan** [Shengqi-Pan/Blog#4]
scikit-learn是一个强大的python机器学习库，建立在Matplotlib和Numpy之上。

# 关于sklearn官方文档

## 官方网文档的内容

![图片](https://user-images.githubusercontent.com/36954327/75445787-13803f00-59a1-11ea-9867-b8d90c5a3044.png)

## 官方文档的结构

![图片](https://user-images.githubusercontent.com/36954327/75445946-62c66f80-59a1-11ea-9e51-6bbf7c57497b.png)
由图可知，该库的算法主要有四类：分类、回归、聚类和降维。

- 常用的回归：线性、决策树、SVM、KNN ；集成回归：随机森林、Adaboost、GradientBoosting、Bagging、ExtraTrees
- 常用的分类：线性、决策树、SVM、KNN，朴素贝叶斯；集成分类：随机森林、Adaboost、GradientBoosting、Bagging、ExtraTrees
- 常用聚类：k均值（K-means）、层次聚类（Hierarchical clustering）、DBSCAN
- 常用降维：LinearDiscriminantAnalysis、PCA

这个流程图中：蓝色圆圈是判断条件，绿色方框是可以选择的算法，我们可以根据自己的数据特征和任务目标去找一条自己的操作路线。
  sklearn的六大板块如下图所示
  ![图片](https://user-images.githubusercontent.com/36954327/75446195-eaac7980-59a1-11ea-9d9b-c7aefc207d04.png)

# sklearn快速食用

## 传统机器学习任务的步骤

传统机器学习的一般流程是：获取数据->数据预处理->训练模型->模型评估->预测、分类

##第一步：获取数据
### 从sklearn导入数据集
sklearn自带有玩具数据集和真实世界的数据集如下

![图片](https://user-images.githubusercontent.com/36954327/75521462-c785d680-5a42-11ea-89ab-a830bd4349a2.png)

![图片](https://user-images.githubusercontent.com/36954327/75521476-d2d90200-5a42-11ea-843c-240c9a5275fe.png)
> 需要注意的是：玩具数据集虽然有助于快速说明sklearn中各种算法的行为，但数据规模往往太小，无法代表真实世界的机器学习任务

> 具体使用可参考：https://sklearn.apachecn.org/docs/0.21.3/47.html
### 创建随机数据集

略

### 从外部导入数据集

> 以下摘自sklearn中文文档

scikit-learn使用任何存储为numpy数组或者scipy稀疏数组的数值数据。 其他可以转化成数值数组的类型也可以接受，如pandas中的DataFrame。

以下推荐一些将标准纵列形式的数据转换为scikit-learn可以使用的格式的方法:

- pandas.io 提供了从常见格式(包括CSV,Excel,JSON,SQL等)中读取数据的工具.DateFrame 也可以从由 元组或者字典组成的列表构建而成.Pandas能顺利的处理异构的数据，并且提供了处理和转换 成方便scikit-learn使用的数值数据的工具。
- scipy.io 专门处理科学计算领域经常使用的二进制格式，例如.mat和.arff格式的内容。
- numpy/routines.io 将纵列形式的数据标准的加载为numpy数组
- scikit-learn的datasets.load_svmlight_file处理svmlight或者libSVM稀疏矩阵
- scikit-learn的 datasets.load_files 处理文本文件组成的目录，每个目录名是每个 类别的名称，每个目录内的每个文件对应该类别的一个样本

对于一些杂项数据，例如图像，视屏，音频。您可以参考:

- skimage.io 或 Imageio 将图像或者视频加载为numpy数组
- scipy.misc.imread (requires the Pillow package)将各种图像文件格式加载为 像素灰度数据
- scipy.io.wavfile.read 将WAV文件读入一个numpy数组

存储为字符串的无序(或者名字)特征(在pandas的DataFrame中很常见)需要转换为整数，当整数类别变量 被编码成独热变量(sklearn.preprocessing.OneHotEncoder)或类似数据时，它或许可以被最好的利用。 参见预处理数据.

注意：如果你要管理你的数值数据，建议使用优化后的文件格式来减少数据加载时间,例如HDF5。像 H5Py, PyTables和pandas等的各种库提供了一个Python接口，来读写该格式的数据。

## 第二步：数据预处理

数据预处理阶段是机器学习中不可缺少的一环，它会使得数据更加有效的被模型或者评估器识别。sklearn.preprocessing包提供了几个常见的实用功能和变换器类型，用来将原视特征相邻更改为更适合机器学习模型的形式。
常见有以下几种：
- StandardScaler
- MaxAbsScaler
- MinMaxScaler
- RobustScaler
- Normalizer

### 数据标准化

数据集的*标准化*对scikit-learn中实现的大多数机器学习算法来说是*常见的要求*。如果个别特征或多或少看起来不是很像标准正态分布(*具有零均值和单位方差*)，那么它们的表现力可能会较差。

经过数据标准化变换后各维特征有0均值，单位方差，也叫z-score规范化（零均值规范化），计算方式是将特征值减去均值，除以标准差。

### 最小-最大规范化

将样本进行线性变换到[0,1]或其他区间

### 正则化

计算样本的p范数，然后将样本的所有元素都除以该范数。

### one-hot编码

one-hot编码是一种对离散特征值的编码方式，在LR模型中常用到，用于给线性模型增加非线性能力。

### 特征二值化

给定阈值，将特征转换为0/1。

```python
binarizer = sklearn.preprocessing.Binarizer(threshold=1.1)
binarizer.transform(X)
```

## 第三步：数据集拆分

```python
from sklearn.mode_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
```

## 第四步：定义与训练模型

直接见https://www.cnblogs.com/wj-1314/p/10179741.html

## 第六步：保存、预测

参考资料：
https://www.cnblogs.com/wj-1314/p/10179741.html
https://scikit-learn.org/stable/modules/classes.html
